### DISCUSSION TITLE
Integrating TrueType font support (I have the code, are you interested?)

### DISCUSSION THREAD

**Discussion by codewitch-honey-crisis** [Bodmer/TFT_eSPI#2020] [General]
I recently attempted to integrate my Tiny_TTF library into LVGL. It's integrated pending some build script issues on their end that they can't solve so I've closed the pull request. I'm a bit frustrated by it, but I ramble.

Anyway, days of effort are currently in question because of it, but I figured when life gives you lemons, make lemonade.

Point is I've ported this https://github.com/nothings/stb/blob/master/stb_truetype.h to work in low memory environments by streaming everything, and tweaking the heap scheme a little bit.

It's tested well enough that I use it in [htcw_gfx](https://honeythecodewitch.com/gfx). 

Before I go through the effort of integrating, I was wondering if you're even interested in incorporating it.

Here's my code. It's a header only C library which serves as its own source file so it's easy to incorporate as a private header
You would include it in TFT_eSPI.cpp or similar, maybe it's own translation unit, it's really up to you. https://github.com/codewitch-honey-crisis/stb/blob/master/stb_truetype.h

**Comment by Bodmer** [comment]
Thanks for the offer. At the moment quite a few different font formats are supported and I am reluctant to add anything new, but I will add to the potential enhancement list.

**Comment by Bodmer** [comment]
This may be of interest. There is a library here:
https://github.com/takkaO/OpenFontRender
That supports rendering of TrueType fonts. Due to the complexity of the code required to convert the font format to a pixel map this is quite slow to render but may be of interest for static text.

**Comment by Bodmer** [comment]
If your TTF rendering code could be incorporated into an independant library like OpenFontRender with TFT_eSPI compatible examples then it would be an interesting addition. I think your code looks like it will render faster.

**Comment by codewitch-honey-crisis** [comment]
It's possible. Right now I have a 3rd party lib I created for LVGL that uses it. I contributed to that project, but I'm sure I could adapt it.

**Reply by Bodmer** [reply]
I tried to get tiny_ttf to work with TFT_eSPI but failed! I may have another go at some point.

**Reply by codewitch-honey-crisis** [reply]
It looks like it may take some doing. I feel like I could adapt the GFXFont code if I understood it better. Understanding it better would actually help me on the backend too because I'd like to have support for it in my own offering as well.

as an aside, I noticed you implemented the writing half of the ESP32-S3 parallel i8080 interface. I guess you ran into the same problem I did reading the pins back. I think we might be able to use GDMA for that, but it's scarcely documented and I haven't gotten far with it. In the alternative I'm creating a widget library such that I can render parts of the display on demand using partial bitmaps and DMA (where supported). I hate to do it, since LVGL already offers this, but short of one of us cracking the GDMA situation and fixing 3-wire SPI/SDA Reads on the latest Arduino frameworks on the ESP32 line it feels like we're being corralled in that direction. I've raised the issue with espressif, and even tried to help them fix it to a degree but ran into a wall.

**Reply by Bodmer** [reply]
OK, thanks for looking at this. Sounds like there are too many blocking issues to progress this.
